Add missing GC.SuppressFinalize

adamsitnik opened the pull request:
fixes #64741

adamsitnik committed:
add tests (they are currently failing)

adamsitnik committed:
remove redundant call to base.Dispose

adamsitnik committed:
fix the bug

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IO/Strategies/BufferedFileStreamStrategy.cs:
```diff
@@ -137,7 +137,15 @@ public override async ValueTask DisposeAsync()
+            {
+                GC.SuppressFinalize(this);
+            }
+        }
```
Do we need DisposeInternal at all?  Could FileStream's Dispose(bool) instead do:
```C#
if (disposing) _strategy.Dispose();
```
?

adamsitnik replied:
@stephentoub to answer your question I've written some additional tests (to cover all scenarios). And I've discovered a new bug: for types that derive from `FileStream` and have buffering enabled, the buffer is currently not getting flushed when the instance is finalized. Repro:

```cs
public class DerivedFileStream : FileStream
{
    public DerivedFileStream(string path, FileMode mode, FileAccess access, FileShare share, int bufferSize, FileOptions options)
        : base(path, mode, access, share, bufferSize, options)
    {
    }
}

[Fact]
public static void Repro()
{
    string filePath = Path.Combine(Path.GetTempPath(), Path.GetTempFileName());
    Separate(filePath);
    
    GC.Collect();
    GC.WaitForPendingFinalizers();
    
    Assert.Equal(new byte [] { 1} , File.ReadAllBytes(filePath));

    // use a separate method to be sure that fs isn't rooted at time of GC.
    static void Separate(string filePath)
    {
        FileStream fs = new DerivedFileStream(filePath, FileMode.Create, FileAccess.Write, FileShare.None, 4096, FileOptions.None);
        fs.WriteByte(1);
    }
}
```

It's caused by the fact that `~DerivedFileStreamStrategy` is currently passing `false` to `Dispose`:

https://github.com/dotnet/runtime/blob/4f2cd55e7c0feeff795b8b43ab215f43ebba3245/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/DerivedFileStreamStrategy.cs#L26-L31

which is perfectly fine, because derived type is expecting that.

However, when it's passed back to `BufferedFileStreamStrategy.Dispose` the `Flush` is not performed for `!disposing`:

https://github.com/dotnet/runtime/blob/4f2cd55e7c0feeff795b8b43ab215f43ebba3245/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/BufferedFileStreamStrategy.cs#L142-L150

`~BufferedFileStreamStrategy` workarounds it by always passing `true` to `Dispose`:

https://github.com/dotnet/runtime/blob/4f2cd55e7c0feeff795b8b43ab215f43ebba3245/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/BufferedFileStreamStrategy.cs#L34-L40

My current take on it is that removing finalizer from `FileStream` was a bad idea as it actually got things more complicated and regressed the performance instead of improving it. I think that we should restore it, let it run for a few weeks on main and backport it to 6.0. PTAL at my recent changes.

adamsitnik committed:
add test for derived FileStream with a finalizer

adamsitnik committed:
add test for derived FileStream without a finalizer

adamsitnik committed:
restore finalizer to the FileStream

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs:
```diff
@@ -49,6 +49,11 @@ public FileStream(IntPtr handle, FileAccess access, bool ownsHandle, int bufferS
+                if (!_strategy.RequiresFinalizer)
+                {
+                    GC.SuppressFinalize(this);
+                }
```
Do we really need to do this?  Do we need to be bending over backwards to try to improve the performance of code that neglects to dispose of a FileStream?

adamsitnik replied:
we don't, I am going to revert this change

adamsitnik committed:
address code review feedback

rpnetcoder commented:
Dear Team, this is just a quick follow-up.

As the author of original issue, I would like to emphasize the significance of the bug. In my particular case, there are hundreds of thousands of FileStream objects created daily which are properly disposed after being used, but not GC collected in Gen 0 and moved to the Finalization queue.  I can imagine even much larger services which might use FileStream under the hood like Microsoft Azure, where, if this bug gets fixed, the whole .NET ecosystem will benefit from it. 

Thanks

adamsitnik commented:
@rpnetcoder thank you for your feedback! I am definitely willing to fix the bug and do my best to backport it to 6.0.

@stephentoub could you please take a look at the PR? I've addressed your feedback

adamsitnik commented:
/backport to release/6.0
